### PR TITLE
Fixed #30328 -- Fixed crash of IntegerField.validators when limit_value in a custom validator is callable.

### PR DESCRIPTION
**Pull request by scottfits**
The field validators MaxValueValidator and MinValueValidator are compared to the limits enforced by MySQL on IntegerField size to see if they should be added to a field's validators. Since 24cae0bedc51093b363c323af555946a8edea1a1 allowed limit_value to be a function, the comparison was raising a TypeError because it was compared to an integer.

### CODE REVIEW

File: django/core/validators.py [hunk 1]
```diff
@@ -311,25 +311,28 @@ class BaseValidator:
...
     def __eq__(self, other):
         return (
             isinstance(other, self.__class__) and
-            self.limit_value == other.limit_value and
+            self.get_limit_value() == other.get_limit_value() and
             self.message == other.message and
             self.code == other.code
         )
 
+    def get_limit_value(self):
```
Comment (pope1ni) [inline django/core/validators.py]: I'd rename this to `limit_value` and make it a `@property`...

File: django/core/validators.py [hunk 2]
```diff
@@ -311,25 +311,28 @@ class BaseValidator:
...
+        limit_value = self.get_limit_value()
         params = {'limit_value': limit_value, 'show_value': cleaned, 'value': value}
         if self.compare(cleaned, limit_value):
             raise ValidationError(self.message, code=self.code, params=params)
 
     def __eq__(self, other):
         return (
             isinstance(other, self.__class__) and
-            self.limit_value == other.limit_value and
+            self.get_limit_value() == other.get_limit_value() and
```
Comment (pope1ni) [inline django/core/validators.py]: Then this change can be reverted.

File: django/db/models/fields/__init__.py [hunk 3]
```diff
@@ -1794,11 +1794,11 @@ def validators(self):
         min_value, max_value = connection.ops.integer_field_range(internal_type)
         if (min_value is not None and not
             any(isinstance(validator, validators.MinValueValidator) and
-                validator.limit_value >= min_value for validator in validators_)):
+                validator.get_limit_value() >= min_value for validator in validators_)):
```
Comment (pope1ni) [inline django/db/models/fields/__init__.py]: And this can be reverted.

File: django/db/models/fields/__init__.py [hunk 4]
```diff
@@ -1794,11 +1794,11 @@ def validators(self):
...
             validators_.append(validators.MinValueValidator(min_value))
         if (max_value is not None and not
             any(isinstance(validator, validators.MaxValueValidator) and
-                validator.limit_value <= max_value for validator in validators_)):
+                validator.get_limit_value() <= max_value for validator in validators_)):
```
Comment (pope1ni) [inline django/db/models/fields/__init__.py]: And this can be reverted.

File: django/core/validators.py [hunk 5]
```diff
@@ -311,25 +311,28 @@ class BaseValidator:
...
     def __init__(self, limit_value, message=None):
-        self.limit_value = limit_value
+        self._limit_value = limit_value
         if message:
             self.message = message
 
     def __call__(self, value):
         cleaned = self.clean(value)
-        limit_value = self.limit_value() if callable(self.limit_value) else self.limit_value
+        limit_value = self.get_limit_value()
```
Comment (pope1ni) [inline django/core/validators.py]: This can be changed to use the new property.

**Comment by scottfits** [comment]
Thanks @pope1ni for feedback. I am happy to provide test but not sure how given the test is dependent on the mysql connection. Should i just mock the connection and put it in tests/validators?

I think an example of the this regression is here where the CI fails, but it was force merged anyway. Can we re-run the CI for this PR?
https://github.com/django/django/pull/11163

File: docs/releases/2.2.1.txt [hunk 6]
```diff
@@ -18,7 +18,10 @@ Bugfixes
 * Fixed a regression in Django 2.2 that caused a crash when loading the
   template for the technical 500 debug page (:ticket:`30324`).
 
+* Fixed a regression in Django 2.2 that caused a crash when IntegerField
+  limits are callable.
```
Comment (pope1ni) [inline docs/releases/2.2.1.txt]: Add `` (:ticket:`30328`)`` to the end before the `.` and move this entry to the end.

File: docs/releases/2.2.1.txt [hunk 7]
```diff
@@ -18,7 +18,10 @@ Bugfixes
...
+
 * Fixed crash of ``ordering`` argument in
   :class:`~django.contrib.postgres.aggregates.ArrayAgg` and
   :class:`~django.contrib.postgres.aggregates.StringAgg` when it contains an
-  expression with params (:ticket:`30332`).
+  expression with params (:ticket:`30332`).
```
Comment (pope1ni) [inline docs/releases/2.2.1.txt]: Please ensure that there is a new line at the end of the file, i.e. revert the change that removes it.

File: django/core/validators.py [hunk 8]
```diff
@@ -330,6 +329,10 @@ def __eq__(self, other):
             self.code == other.code
         )
 
+    @property
+    def limit_value(self):
+        return self._limit_value() if callable(self._limit_value) else self._limit_value
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/core/validators.py]: Or make it a settable property then. I figured it wasn't common to change it in that manner, but it is easy to fix.

Regarding `__eq__`, I thought about that too, and it could just compare `_limit_value`.

I worry about other code added in future (or third-party code) not being aware that it needs to handle callables.

[the remark this answers is not in the document]
Comment (felixxm) [reply django/core/validators.py]: In Django itself `limit_value` is used only in `IntegerField.validators` and due to the 3rd party apps we added a note in release notes and changed documentation so I would not worry about them.

We want to backport this fix to the `2.2` so IMO we should avoid unnecessary refactoring which may cause next regressions.

[the remark this answers is not in the document]
Comment (pope1ni) [reply django/core/validators.py]: Fair enough. I think the change I've suggested should be considered for master then as it'll be more robust in the long run.

**Comment by felixxm** [comment]
@scottfits Are you able to address remaining comments in this week? If not please let me know :+1: , I can finish this for you :pill: .

**Comment by scottfits** [comment]
Yes I can handle it - should I revert limit_value to no longer be a property and simply check for its type in the validator?

**Comment by felixxm** [comment]
@scottfits Yes, this sounds good (see [comment](https://github.com/django/django/pull/11180#discussion_r272846270)).

**Comment by scottfits** [comment]
Because the min and max limits only exist with mysql and oracle backends, i tested this by using a custom test settings file with the mysql backend enabled.

When running my test test_callable_limit_value_validators, the output confirms the regression

```
  File "/Users/scott/opensource/django/django/db/models/fields/__init__.py", line 1797, in <genexpr>
    validator.limit_value >= min_value for validator in validators_)):
TypeError: '>=' not supported between instances of 'function' and 'int'
```

After the patch, the test passes.

**Comment by felixxm** [comment]
@scottfits Fix is missing :thinking: , please update also release notes.

**Comment by felixxm** [comment]
@scottfits Thanks again for the patch :+1: I squashed commits, rebased from master, fixed release notes and missing parentheses.
